acstools/acscteforwardmodel.py

```diff
@@ -0,0 +1,254 @@
+
+In Pyraf::
+
+    --> import acstools
+    --> epar acscteforwardmodel
+
+For help usage use ``exe_args=['--help']``
+
+"""
+# STDLIB
+import os
+import subprocess
+
+__taskname__ = "acscteforwardmodel"
+__version__ = "1.0"
+__vdate__ = "19-Jul-2018"
+__all__ = ['acscteforwardmodel']
+
+
+def acscteforwardmodel(input, exec_path='', time_stamps=False, verbose=False, 
+               quiet=False, single_core=False, exe_args=None):
+    """
+    Run the acscteforwardmodel.e executable as from the shell.
+
+    Expect input to be ``*_blc_tmp.fits`` or ``*_flc.fits``.
+    Output is automatically named ``*_ctefmod.fits``.
+
+    Parameters
+    ----------
+    input : str or list of str
+        Input filenames in one of these formats:
+
+            * a single filename ('j1234567q_blc_tmp.fits')
+            * a Python list of filenames
+            * a partial filename with wildcards ('\*blc_tmp.fits')
+            * filename of an ASN table ('j12345670_asn.fits')
+            * an at-file (``@input``)
+
+    exec_path : str, optional
+        The complete path to ACSCTE forward model executable.
+        If not given, run ACSCTE given by 'acscteforwardmodel.e'.
+
+    time_stamps : bool, optional
+        Set to True to turn on the printing of time stamps.
+
+    verbose : bool, optional
+        Set to True for verbose output.
+
+    quiet : bool, optional
+        Set to True for quiet output.
+
+    single_core : bool, optional
+        CTE correction in the ACSCTE forward model will by default try to use
+        all available CPUs on your computer. Set this to True to force the use
+        of just one CPU.
+
+    exe_args : list, optional
+        Arbitrary arguments passed to underlying executable call.
+        Note: Implementation uses subprocess.call and whitespace is not
+        permitted. E.g. use exe_args=['--nThreads', '1']
+
+    """
+    from stsci.tools import parseinput  # Optional package dependency
+
+    if exec_path:
+        if not os.path.exists(exec_path):
+            raise OSError('Executable not found: ' + exec_path)
+        call_list = [exec_path]
+    else:
+        call_list = ['acscteforwardmodel.e']
+
+    # Parse input to get list of filenames to process.
+    # acscte.e only takes 'file1,file2,...'
+    infiles, dummy_out = parseinput.parseinput(input)
+
+    # Make temporary files for input into forward model
+    tmp_files = make_tmp_files(infiles)
+
+    call_list.append(','.join(tmp_files))
+
+    if time_stamps:
+        call_list.append('-t')
+
+    if verbose:
+        call_list.append('-v')
+
+    if quiet:
+        call_list.append('-q')
+
+    if single_core:
+        call_list.append('-1')
+
+    if exe_args:
+        call_list.extend(exe_args)
+
+    subprocess.check_call(call_list)
+
+    # Restore extra FLC extensions to forward model
+    restore_files(infiles)
+
+
+def make_tmp_files(infiles):
+    """
+    If more than 6 extensions, take SCI, ERR, and DQ extensions and make a
```

Comment:
Is this even what is needed? I didn't think the issue was specific to 6, isn't the 6 just an artifact of the input file that was used? That after the 1st 3 imsets (6 exts) that file contained other extensions that were not imsets. I think what you actually want is to iterate over the ``hdu.info`` tab and just strip out the full imsets. This is at least what I'm thinking about for the fix in the C code.

Reply:
Yes, you are right that the issue isn't specific to 6. This was simply the first method I thought of that would separate both full-frame and subarray images with the distortion extensions from those without them. Identifying the imsets is probably a better way to go.

Reply:
A shorter workaround would/could be to just copy the entire file to a tmp and then change ``NEXTEND`` (in the new tmp file) to the number of imsets. This would decrease the size of the wrapper.

```python
from astropy.io import fits
hduList = fits.open('<fname>')
nimsets = [hdu.name for hdu in hduList].count('SCI')
```  

However, I would argue that the briefness of this workaround leads to not being included at all, i.e. it remains a workaround and not fix (i.e. it doesn't get merged).

Reply:
I am making an assumption here about the definition of an imset that I am unsure is a vlid one. _c.f._ my question in https://github.com/spacetelescope/hstcal/issues/323#issuecomment-408457258

Reply:
Actually, you'd need ``NEXTEND = nimsets*3`` due to [this ugly line of code](https://github.com/spacetelescope/hstcal/blob/2.0.0/pkg/acs/calacs/lib/getacskeys.c#L95):
```c
...
/* Three extensions per SingleGroup. */
# define EXT_PER_GROUP 3
...
/* Convert number of extensions to number of SingleGroups. */
acs->nimsets = nextend / EXT_PER_GROUP;
...
```

Reply:
_c.f._ https://github.com/spacetelescope/hstcal/issues/323#issuecomment-408524951
```python
nimsets = len(set([hdus.ver for hdus in hduList]))
```